plugins/discovery/discovery_test.go

```diff
@@ -1557,24 +1546,60 @@ func TestStatusUpdates(t *testing.T) {
 	disco.oneShot(ctx, download.Update{ETag: "etag-2"})
 
 	// Check that all updates were received and active revisions are expected.
-	var ok bool
-	var updates []status.UpdateRequestV1
-	t0 := time.Now()
-
-	for !ok && time.Since(t0) < time.Second {
-		updates = ts.Updates()
-		ok = len(updates) == 7 &&
-			updates[0].Plugins["discovery"].State == plugins.StateNotReady && updates[0].Plugins["status"].State == plugins.StateOK &&
-			updates[1].Plugins["discovery"].State == plugins.StateOK && updates[1].Plugins["status"].State == plugins.StateOK &&
-			updates[2].Plugins["discovery"].State == plugins.StateOK && updates[2].Discovery.ActiveRevision == "test-revision-1" && updates[2].Discovery.Code == "" &&
-			updates[3].Plugins["discovery"].State == plugins.StateOK && updates[3].Discovery.ActiveRevision == "test-revision-1" && updates[3].Discovery.Code == "bundle_error" &&
-			updates[4].Plugins["discovery"].State == plugins.StateOK && updates[4].Discovery.ActiveRevision == "test-revision-2" && updates[4].Discovery.Code == "" &&
-			updates[5].Plugins["discovery"].State == plugins.StateOK && updates[5].Discovery.ActiveRevision == "test-revision-2" && updates[5].Discovery.Code == "bundle_error" &&
-			updates[6].Plugins["discovery"].State == plugins.StateOK && updates[6].Discovery.ActiveRevision == "test-revision-2" && updates[6].Discovery.Code == ""
+	expectedDiscoveryUpdates := []struct {
+		Code     string
+		Revision string
+	}{
+		{
+			Code:     "",
+			Revision: "test-revision-1",
+		},
+		{
+			Code:     "bundle_error",
+			Revision: "test-revision-1",
+		},
+		{
+			Code:     "",
+			Revision: "test-revision-2",
+		},
+		{
+			Code:     "bundle_error",
+			Revision: "test-revision-2",
+		},
+		{
+			Code:     "",
+			Revision: "test-revision-2",
+		},
 	}
 
-	if !ok {
-		t.Fatalf("Did not receive expected updates before timeout expired. Received: %+v", updates)
+	// nextExpectedDiscoveryUpdate, we look for each
+	nextExpectedDiscoveryUpdate := expectedDiscoveryUpdates[0]
+	expectedDiscoveryUpdates = expectedDiscoveryUpdates[1:]
+
+	timeout, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	for {
+		select {
+		case update := <-ts.updates:
+			if update.Discovery != nil {
+				matches := false
+				if update.Discovery.Code == nextExpectedDiscoveryUpdate.Code &&
+					update.Discovery.ActiveRevision == nextExpectedDiscoveryUpdate.Revision {
+					matches = true
+				}
+
+				if matches {
```

Comment:
Is it an error if there is no match?

Reply:
Not really, as the next status might be the expected one and then it's fine. This is just reading statuses until the expected one is found, before moving on to the next one.